launcher/launcher.go

```diff
@@ -36,3 +63,805 @@ type NodeInfo struct {
+		ServiceID: instance.ServiceID, SubjectID: instance.SubjectID,
+		Instance: index,
+	}, Priority: instance.Priority}
+
+	uid, err := launcher.storage.GetInstanceUID(instanceInfo.InstanceIdent)
+	if err != nil {
+		if !errors.Is(err, ErrNotExist) {
+			log.Errorf("Can't get instance UID: %v", err)
+			return instanceInfo, aoserrors.Errorf("Can't get instance UID: %v", err)
+		}
+
+		uid, err = launcher.uidPool.GetFreeID()
+		if err != nil {
+			log.Errorf("Can't get free UID: %v", err)
+			return instanceInfo, aoserrors.Errorf("Can't get free UID: %v", err)
+		}
+	}
+
+	instanceInfo.UID = uint32(uid)
+
+	stateStrageParams := storagestate.SetupParams{
+		InstanceIdent: instanceInfo.InstanceIdent,
+		UID:           uid, GID: int(service.GID),
+	}
+
+	if service.Config.Quotas.StateLimit != nil {
+		stateStrageParams.StateQuota = *service.Config.Quotas.StateLimit
+	}
+
+	if service.Config.Quotas.StorageLimit != nil {
+		stateStrageParams.StorageQuota = *service.Config.Quotas.StorageLimit
+	}
+
+	instanceInfo.StoragePath, instanceInfo.StatePath, err = launcher.storageStateProvider.Setup(stateStrageParams)
+	if err != nil {
+		log.Errorf("Can't setup storage and state for instance: %v", err)
+
+		_ = launcher.uidPool.RemoveID(uid)
+
+		return instanceInfo, aoserrors.Errorf("Can't setup storage and state for instance: %v", err)
+	}
+
+	return instanceInfo, nil
+}
+
+func (launcher *Launcher) getNodesByStaticResources(
+	serviceInfo imagemanager.ServiceInfo, instanceInfo cloudprotocol.InstanceInfo,
+) (nodes []*nodeStatus, errStatus cloudprotocol.InstanceStatus) {
+	nodes = launcher.getNodeByRunner(serviceInfo.Config.Runner)
+	if len(nodes) == 0 {
+		log.Errorf("No appropriate node with runner: %s", serviceInfo.Config.Runner)
+
+		return nodes, createInstanceStatusFromInfo(instanceInfo, 0, serviceInfo.AosVersion,
+			cloudprotocol.InstanceStateFailed, fmt.Sprintf(
+				"No appropriate node with runner: %s", serviceInfo.Config.Runner))
+	}
+
+	nodes = launcher.getNodesByResources(nodes, serviceInfo.Config.Resources)
+	if len(nodes) == 0 {
+		log.Errorf("No appropriate node with resources: %v", serviceInfo.Config.Resources)
+
+		return nodes, createInstanceStatusFromInfo(instanceInfo, 0, serviceInfo.AosVersion,
+			cloudprotocol.InstanceStateFailed, fmt.Sprintf(
+				"No appropriate node with resourceser: %v", serviceInfo.Config.Resources))
+	}
+
+	nodes = launcher.getNodesByLabels(nodes, instanceInfo.Labels)
+	if len(nodes) == 0 {
+		log.Errorf("No appropriate node with labels %v", instanceInfo.Labels)
+
+		return nodes, createInstanceStatusFromInfo(instanceInfo, 0, serviceInfo.AosVersion,
+			cloudprotocol.InstanceStateFailed, fmt.Sprintf(
+				"No appropriate node with labels: %v", instanceInfo.Labels))
+	}
+
+	return nodes, errStatus
+}
+
+func (launcher *Launcher) getNodesByDevices(
+	nodes []*nodeStatus, desiredDevices []aostypes.ServiceDevice,
+) (newNodes []*nodeStatus) {
+	if len(desiredDevices) == 0 {
+		return slices.Clone(nodes)
```

Comment:
why do we need clone if we could return just `nodes`

Reply:
in the next methods, we will modify result slice but original one should stay original